storage/internal/benchmarks/main.go

```diff
@@ -269,96 +286,33 @@ func main() {
 	closePools := initializeClientPools(ctx, opts)
 	defer closePools()
 
-	w := os.Stdout
-
-	if outputFile != "" {
-		w = file
-		// The output file is only for benchmarking data points; if sending
-		// output to a file, we can use stdout for informational logs
-		fmt.Printf("Benchmarking started: %s\n", start.UTC().Format(time.ANSIC))
-		fmt.Printf("Code version: %s\n", codeVersion)
-		fmt.Printf("Results file: %s\n", outputFile)
-		fmt.Printf("Bucket:  %s\n", opts.bucket)
-		fmt.Printf("Benchmarking options: %+v\n", opts)
-	}
-
 	if err := populateDependencyVersions(); err != nil {
 		log.Fatalf("populateDependencyVersions: %v", err)
 	}
 
 	if err := warmupW1R3(ctx, opts); err != nil {
 		log.Fatal(err)
 	}
-	recordResultGroup, _ := errgroup.WithContext(ctx)
-	startRecordingResults(w, recordResultGroup, opts.outType)
-
-	simultaneousGoroutines := opts.numWorkers
-
-	// Directories parallelize on the object level, so only run one benchmark at a time
-	if opts.workload == 6 {
-		simultaneousGoroutines = 1
-	}
-
-	benchGroup, ctx := errgroup.WithContext(ctx)
-	benchGroup.SetLimit(simultaneousGoroutines)
-
-	exitWithErrorCode := false
 
 	if opts.enableTracing {
 		cleanup := enableTracing(ctx, opts.traceSampleRate)
 		defer cleanup()
 	}
 
-	// Run benchmarks
-	for i := 0; i < opts.numSamples && time.Since(start) < opts.timeout; i++ {
-		benchGroup.Go(func() error {
-			var benchmark benchmark
-			benchmark = &w1r3{opts: opts, bucketName: opts.bucket}
-
-			if opts.workload == 6 {
-				benchmark = &directoryBenchmark{opts: opts, bucketName: opts.bucket, numWorkers: opts.numWorkers}
-			}
-
-			if err := benchmark.setup(ctx); err != nil {
-				// If setup failed once, it will probably continue failing.
-				// Returning the error here will cancel the context to stop the
-				// benchmarking.
-				return fmt.Errorf("run setup failed: %v", err)
-			}
-			if err := benchmark.run(ctx); err != nil {
-				// If a run fails, we continue, as it could be a temporary issue.
-				// We log the error and make sure the program exits with an error
-				// to indicate that we did see an error, even though we continue.
-				log.Printf("run failed: %v", err)
-				exitWithErrorCode = true
-			}
-			if err := benchmark.cleanup(); err != nil {
-				// If cleanup fails, we continue, as a single fail is not critical.
-				// We log the error and make sure the program exits with an error
-				// to indicate that we did see an error, even though we continue.
-				// Cleanup may be expected to fail if there is an issue with the run.
-				log.Printf("run cleanup failed: %v", err)
-				exitWithErrorCode = true
-			}
-			return nil
-		})
+	if serverMode {
+		// Server mode blocks forever servicing probe requests
+		runInServerMode(ctx, opts)
```

Comment:
Should this have an exit()? If runInServerMode does return then we would probably not want to continue right?

Reply:
That is correct; although it should never return, I've added a check here to make sure we do error out if it does.